fms/utils/generation.py

```diff
@@ -74,9 +174,33 @@ def generate(
     kwargs: MutableMapping[str, Any] = dict()
     kwargs["past_key_value_states"] = None
     kwargs["use_cache"] = use_cache
+    kwargs["position_ids"] = position_ids
 
-    for _ in range(max_new_tokens):
+    for i in range(max_new_tokens):
         input_ids = next_input[:, -max_seq_len:]
+
+        # prefill
+        if i == 0:
+            kwargs["mask"] = __create_prefill_mask(
+                model_input_lengths, input_ids.device
+            )
+        # decode
+        else:
+            # add 1 for generate
+            model_input_lengths = [
+                model_input_lengths + 1 for model_input_lengths in model_input_lengths
+            ]
+            kwargs["mask"] = __create_decode_mask(model_input_lengths, input_ids.device)
```

Comment:
do all models accept 'mask' kwargs? we could check `mask in inspect.getfullargspec(model.forward).args` to be able to raise a proper error message explicitly.

Reply:
just checked, they all accept `mask` and `position_ids`

Reply:
also, none of this extra mask handling code is necessary in the case where input_ids was a tensor, right? if we were to add an error message, the workarounds it would suggest might not actually be possible in the current implementation.

Reply:
> just checked, they all accept mask and position_ids

there's an alternate generate function in extras, but those paged models don't accept position_ids. I think it's good for this function to continue to work with models that don't accept masks or pos IDs. New models also might not want to accept masks (when nestedtensor compiles), and I'm not sure if position_ids are needed in mamba

Reply:
@JRosenkranz did we change anything that will allow this to (continue to) work in cases where a user has unbatched input to a model that doesn't accept mask+position_ids?

Reply:
Ah I thought that was going to maybe be handled in a separate PR https://github.com/foundation-model-stack/foundation-model-stack/pull/253#discussion_r1661377902

Reply:
nicer error messages can come later if there are cases that always would have failed, but I think we should continue to support cases we already supported. The original generate function worked with basically any model taking a tensor of inputs as a first positional arg (and optionally a past_key_value_states).